File: packages/typescript/sdk/core/src/validate-params.ts
import type { JsonSchema } from "./types";

/**
 * Minimal JSON Schema validator for affordance invoke params.
 *
 * Covers the subset the SLOP affordance schema uses in practice:
 * - `type`: "object" | "string" | "number" | "integer" | "boolean" | "array" | "null"
 * - `required`: string[]   (for objects)
 * - `properties`: Record<string, JsonSchema>  (for objects)
 * - `items`: JsonSchema    (for arrays)
 * - `enum`: readonly unknown[]
 *
 * Returns `null` on success, or a human-readable error message describing the
 * first failure. The message is used as the `message` field of an
 * `invalid_params` error result — see spec/core/affordances.md §invocation.
 */
export function validateParams(schema: JsonSchema | undefined, params: unknown): string | null {
  if (!schema) return null;
  return validate(schema, params, "params");
}

function validate(schema: JsonSchema, value: unknown, path: string): string | null {
  if (schema.enum && !schema.enum.some((v) => deepEqual(v, value))) {
    return `${path} must be one of ${JSON.stringify(schema.enum)}`;
  }

  switch (schema.type) {
    case "object": {
      if (value === null || typeof value !== "object" || Array.isArray(value)) {
        return `${path} must be an object`;
      }
      const obj = value as Record<string, unknown>;
      for (const key of schema.required ?? []) {
        if (!(key in obj)) {
          return `${path}.${key} is required`;
        }
      }
      if (schema.properties) {
        for (const [key, propSchema] of Object.entries(schema.properties)) {
          if (key in obj) {
            const err = validate(propSchema, obj[key], `${path}.${key}`);
            if (err) return err;
          }
        }
      }
      return null;
    }
    case "array": {
      if (!Array.isArray(value)) return `${path} must be an array`;
      if (schema.items) {
        for (let i = 0; i < value.length; i++) {
          const err = validate(schema.items, value[i], `${path}[${i}]`);
          if (err) return err;
        }
      }
      return null;
    }
    case "string":
      return typeof value === "string" ? null : `${path} must be a string`;
    case "number":
      return typeof value === "number" && Number.isFinite(value) ? null : `${path} must be a number`;
    case "integer":
      return typeof value === "number" && Number.isInteger(value) ? null : `${path} must be an integer`;
    case "boolean":
      return typeof value === "boolean" ? null : `${path} must be a boolean`;
    case "null":
      return value === null ? null : `${path} must be null`;
    default:
      // Unknown type — be permissive rather than reject.
      return null;
  }
}

function deepEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (typeof a !== typeof b) return false;
  if (a === null || b === null) return false;
  if (typeof a !== "object") return false;

  const aIsArray = Array.isArray(a);
  const bIsArray = Array.isArray(b);
  if (aIsArray !== bIsArray) return false;

  if (aIsArray) {
    const arrB = b as unknown[];
    if (a.length !== arrB.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (!deepEqual(a[i], arrB[i])) return false;
    }
    return true;
  }

  // Object: compare by key set, not by serialized order. Python/Go/Rust
  // validators treat {a,b} and {b,a} as equal; TypeScript must too, otherwise
  // enum deep-equality rejects semantically identical JSON across SDKs.
  const objA = a as Record<string, unknown>;
  const objB = b as Record<string, unknown>;
  const keysA = Object.keys(objA);
  const keysB = Object.keys(objB);
  if (keysA.length !== keysB.length) return false;
  for (const k of keysA) {
    if (!Object.hasOwn(objB, k)) return false;
    if (!deepEqual(objA[k], objB[k])) return false;
  }
  return true;
}

File: packages/python/slop-ai/src/slop_ai/validate_params.py
"""Minimal JSON Schema validator for affordance invoke params.

Mirrors the TypeScript `validateParams` so the `invalid_params` error code is
reliable across SDKs. Covers the subset the SLOP affordance schema uses:

* ``type``: object | string | number | integer | boolean | array | null
* ``required``: list[str]                  (for objects)
* ``properties``: dict[str, schema]         (for objects)
* ``items``: schema                         (for arrays)
* ``enum``: list[Any]
"""

from __future__ import annotations

import json
from typing import Any


def validate_params(schema: dict[str, Any] | None, params: Any) -> str | None:
    """Return ``None`` on success, or a human-readable error message."""
    if not schema:
        return None
    return _validate(schema, params, "params")


def _validate(schema: dict[str, Any], value: Any, path: str) -> str | None:
    if "enum" in schema and not any(_deep_equal(option, value) for option in schema["enum"]):
        return f"{path} must be one of {json.dumps(schema['enum'])}"

    t = schema.get("type")
    if t == "object":
        if not isinstance(value, dict):
            return f"{path} must be an object"
        for key in schema.get("required", []) or []:
            if key not in value:
                return f"{path}.{key} is required"
        for key, prop_schema in (schema.get("properties") or {}).items():
            if key in value:
                err = _validate(prop_schema, value[key], f"{path}.{key}")
                if err:
                    return err
        return None
    if t == "array":
        if not isinstance(value, list):
            return f"{path} must be an array"
        items = schema.get("items")
        if items:
            for i, item in enumerate(value):
                err = _validate(items, item, f"{path}[{i}]")
                if err:
                    return err
        return None
    if t == "string":
        return None if isinstance(value, str) else f"{path} must be a string"
    if t == "number":
        return None if isinstance(value, (int, float)) and not isinstance(value, bool) else f"{path} must be a number"
    if t == "integer":
        return None if isinstance(value, int) and not isinstance(value, bool) else f"{path} must be an integer"
    if t == "boolean":
        return None if isinstance(value, bool) else f"{path} must be a boolean"
    if t == "null":
        return None if value is None else f"{path} must be null"
    # Unknown type — be permissive rather than reject.
    return None


def _deep_equal(a: Any, b: Any) -> bool:
    if a is b:
        return True
    if type(a) is not type(b):
        # Allow int/float equivalence.
        if isinstance(a, (int, float)) and isinstance(b, (int, float)):
            return a == b
        return False
    try:
        return json.dumps(a, sort_keys=True) == json.dumps(b, sort_keys=True)
    except TypeError:
        return a == b
